contracts/src/v0.8/functions/dev/v1_X/libraries/ChainSpecificUtil.sol

```diff
@@ -40,18 +40,24 @@ library ChainSpecificUtil {
 
   // ------------ End Optimism Constants ------------
 
-  /// @notice Returns the L1 fees in wei that will be paid for the current transaction, given any calldata
-  /// @notice for the current transaction.
-  /// @notice When on a known Arbitrum chain, it uses ArbGas.getCurrentTxL1GasFees to get the fees.
-  /// @notice On Arbitrum, the provided calldata is not used to calculate the fees.
-  /// @notice On Optimism, the provided calldata is passed to the GasPriceOracle predeploy
-  /// @notice and getL1Fee is called to get the fees.
-  function _getCurrentTxL1GasFees(bytes memory txCallData) internal view returns (uint256 l1FeeWei) {
+  /// @notice Returns the upper limit estimate of the L1 fees in wei that will be paid for L2 chains
+  /// @notice based on the size of the transaction data and the current gas conditions.
+  /// @notice This is an "upper limit" as it assumes the transaction data is uncompressed when posted on L1.
+  function _getL1FeeUpperLimit(uint256 dataSizeBytes) internal view returns (uint256 l1FeeWei) {
     uint256 chainid = block.chainid;
     if (_isArbitrumChainId(chainid)) {
-      return ARBGAS.getCurrentTxL1GasFees();
+      // https://docs.arbitrum.io/build-decentralized-apps/how-to-estimate-gas#where-do-we-get-all-this-information-from
+      (, uint256 l1PricePerByte, , , , ) = ARBGAS.getPricesInWei();
+      return l1PricePerByte * (dataSizeBytes + ARB_L1_FEE_DATA_PADDING_SIZE);
     } else if (_isOptimismChainId(chainid)) {
-      return OVM_GASPRICEORACLE.getL1Fee(bytes.concat(txCallData, L1_FEE_DATA_PADDING));
+      // https://github.com/ethereum-optimism/specs/blob/main/specs/protocol/exec-engine.md#ecotone-l1-cost-fee-changes-eip-4844-da
+      // note we conservatively assume all non-zero bytes: tx_compressed_size = tx_data_size_bytes
+      uint256 l1BaseFeeWei = L1BLOCK.basefee();
+      uint256 l1BaseFeeScalar = L1BLOCK.baseFeeScalar();
+      uint256 l1BlobBaseFeeWei = L1BLOCK.blobBaseFee();
+      uint256 l1BlobBaseFeeScalar = L1BLOCK.blobBaseFeeScalar();
+      uint256 weightedGasPrice = (l1BaseFeeScalar * l1BaseFeeWei + l1BlobBaseFeeScalar * l1BlobBaseFeeWei) / 10 ** 6;
```

Comment:
10^6 is returned in GasPriceOracle as `decimals()`, do we know if OP will always stick to 6 decimals? if not, probably better to read this from GasPriceOracle or have it configurable

Reply:
Thank you for this comment! I will look into this further.

Reply:
Note that we already know we will need to overhaul this logic again in order to support the new L1 fee calculation logic proposed for Fjord, so that might make the hardcoded 10^6 value not that big of a deal. I highly doubt this would change prior to the hardfork.

Context: https://chainlink-core.slack.com/archives/C028EC3TF70/p1715980777258499